### ISSUE TITLE
windows.uninstallTask

### ISSUE THREAD

**Issue by DrusTheAxe** [microsoft/WindowsAppSDK#2779]
### Describe the bug

1

### Steps to reproduce the bug

q

**Comment by riverar** [comment]
> [from rationale] As a package can be removed without the package's direct involvement a mechanism is needed to enable these apps to remove data written to places or ways unknown to the Deployment stack.

[MSIX guarantees a clean uninstall](https://docs.microsoft.com/windows/msix/overview#key-features), so what data are you referring to here? Sounds like if there's untracked data as suggested, this is a MSIX bug.

I'm very concerned this is being requested for a partner to pop open a web browser for some tracking, analytics, exit survey, and/or other generally unwanted behavior.

**Comment by DrusTheAxe** [comment]
> [MSIX guarantees a clean uninstall](https://docs.microsoft.com/windows/msix/overview#key-features), so what data are you referring to here?

Packaged apps can run MediumIL and elevation which can (and do) color outside the lines. For example, some unpackaged apps have extensibility or integration models requiring regkeys or files be written in places **not managed by the Deployment stack**. Apps running at >=MediumIL can write to these locations, but have no reliable mechanism to remove that data when their package is removed.


> Sounds like if there's untracked data as suggested, this is a MSIX bug.

There's a lot of data across Windows (the OS and ecosystem) which isn't managed by MSIX that some apps need to modify. Some of these are impossible for MSIX to properly manage without breaking impact to these unpackaged apps so either there's zero unpackaged apps (only packaged) or we need a way to support these older mechanisms. The former doesn't seem likely anytime soon 😋 so this is a necessity for healthy systems.

> I'm very concerned this is being requested for a partner to pop open a web browser for some tracking, analytics, exit survey, and/or other generally unwanted behavior.

Can a BackgroundTask pop open a web browser to engage in these activities today? That's what windows.preinstallConfigTask and windows.updateTask work today, they activate a manifested background task. Can a background task be defined imperatively (no manifest) and pop open a web browser?...

These uninstall/repair/reset tasks wouldn't enable these actions any more than possible today.

Also, this has been requested by other teams before (via other channels). The need/priority hasn't risen 'above the line' to act on it _so far_, but times and needs can change... This enhancement request is vFuture/backlog, noted to gather feedback and consideration in planning cycles yadda yadda all the usual reasons. Feedback is welcome - on all facets: functionality, restrictions, need, ...

**Comment by riverar** [comment]
The existing `windows.preinstallConfigTask` and `windows.updateTask` tasks seem to be targeting OEMs/MOs (akin to a rescap). Do you know if that's enforced? If so, should this task also have the same restrictions?

**Comment by DrusTheAxe** [comment]
> The existing `windows.preinstallConfigTask` and `windows.updateTask` tasks seem to be targeting OEMs/MOs (akin to a rescap). Do you know if that's enforced?

`windows.preinstallConfigTask` is restricted to **provisioned** (and inbox) packages. OEMs often provision packages but so do others (eg enterprises). Anyone can declare it, but it's only acted on if the package is provisioned (or inbox).

Anyone can provision a package via `PackageManager.ProvisionPackageForAllUsersAsync()`, the Powershell cmdlet `Add-AppxProvisionedPackage` or `dism.exe` **if they have admin privilege**. At that point you p3wn the machine anyway so there's no added threat.

`windows.updateTask` has no such constraint. Any package can declare it and it'll be invoked when a user's registered packaged is updated. No provisioning required, not any capability as I recall.

>If so, should this task also have the same restrictions?

What do you think?

**Comment by dongle-the-gadget** [comment]
I think that there should also be an option for the uninstall action to run as administrator. Either by having something like `RunAsAdmin="true"` or allowing the uninstall app to elevate itself on launch.

As for the restrictions, I feel like `runFullTrust` should be enough.

**Comment by DrusTheAxe** [comment]
https://task.ms/34863092

**Comment by andy-schmidt** [comment]
This looks awesome, we really could use this functionality. This is a definite improvement over using the[ `customInstallActions` restricted capability](https://learn.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-desktop6-uninstallaction). See also https://github.com/MicrosoftDocs/winrt-related/issues/194

What is the status on development of this feature?

**Comment by untoldbyte** [comment]
There is another important use case

Lets say that an msix app regsiters a font to the OS, now msix uninstall will failas the font maybe in use and hence locked

it is possibleto placethe font outside of sandbox and uninstall will succeed but this will keep the fonts registered with no means to unregister

this and similar use cases need custom hooks

**Comment by DrusTheAxe** [comment]
>Lets say that an msix app regsiters a font to the OS

Registers how? Sorry, I'm only lightly savvy about font registration. Is this a manifested registration like `<Extension Category="windows.sharedfont"> or via some Windows API to register a file in the package?

If the latter and the font is loaded and cannot be deleted it shouldn't fail the uninstall -- removing the package _should not fail_. This usually appears as an `ERROR_SHARING_VIOLATION` or `E_ACCESSDENIED` (depending) but remove is, by definition, a best-effort operation. Deployment should not fail an uninstall for such things but rather log the error and move on. Worst case the deployment operation succeeds but the package directory (or something under your ApplicationData directory) remains. In this event the unable-to-be-deleted files/dirs are tracked and 'cleanup' is periodically retried, including worst case after the next reboot.

Partially uninstalling a package won't fail the RemovePackage operation, but having that partial package directory/content hanging around can block a 'new' install of the package (well, it'll seem new since you uninstalled the the package and there's no record of it being present anymore).

But regardless, yes, this is an example where an app can register resources with Windows which the MSIX deployment stack has no knowledge of, and thus no ability to help clean up on uninstall (or update, reset or repair scenarios). `ApplicationData.Local[Cache]Folder` can be a viable option in some cases (copy files to LocalCacheFolder and register there), but only some. Hooks like the uninstall/reset/repair tasks proposed here provide a means to handle those more offbeat scenarios the deployment stack can't.

Manifested features like windows.sharedFonts are another matter entirely. They're designed to be savvy to this issue and shouldn't be a problem. If they are it's a bug and one we'd be very interested to hear about. Please let us know if you find such a thing.

**Comment by untoldbyte** [comment]
> https://task.ms/34863092

cant access

**Comment by untoldbyte** [comment]
the inability to run custom uninstall hook means that my app cant do something simple as creating a user read protected directory to prevent access to proprietary stuff like mov, fonts.

if the folder is set these permissions

the msix uninstall leaves behind the package folder and then it cannot be installed again!

this is a critical issue

**Comment by untoldbyte** [comment]
> >Lets say that an msix app regsiters a font to the OS
> 
> Registers how? Sorry, I'm only lightly savvy about font registration. Is this a manifested registration like `<Extension Category="windows.sharedfont"> or via some Windows API to register a file in the package?
> 
> If the latter and the font is loaded and cannot be deleted it shouldn't fail the uninstall -- removing the package _should not fail_. This usually appears as an `ERROR_SHARING_VIOLATION` or `E_ACCESSDENIED` (depending) but remove is, by definition, a best-effort operation. Deployment should not fail an uninstall for such things but rather log the error and move on. Worst case the deployment operation succeeds but the package directory (or something under your ApplicationData directory) remains. In this event the unable-to-be-deleted files/dirs are tracked and 'cleanup' is periodically retried, including worst case after the next reboot.
> 
> Partially uninstalling a package won't fail the RemovePackage operation, but having that partial package directory/content hanging around can block a 'new' install of the package (well, it'll seem new since you uninstalled the the package and there's no record of it being present anymore).
> 
> But regardless, yes, this is an example where an app can register resources with Windows which the MSIX deployment stack has no knowledge of, and thus no ability to help clean up on uninstall (or update, reset or repair scenarios). `ApplicationData.Local[Cache]Folder` can be a viable option in some cases (copy files to LocalCacheFolder and register there), but only some. Hooks like the uninstall/reset/repair tasks proposed here provide a means to handle those more offbeat scenarios the deployment stack can't.
> 
> Manifested features like windows.sharedFonts are another matter entirely. They're designed to be savvy to this issue and shouldn't be a problem. If they are it's a bug and one we'd be very interested to hear about. Please let us know if you find such a thing.

we need to make a directory read protected by applying special permissions using ACL or similar. I have tried and the uninstall leaves behind the whole package folder i.e. appname+hash inside packages. 

will this get auto deleted on reboot m sure not cause the permissions are still there.



any guidance on how we can achieve this would be great.

also no matter what, lets say if we need to unregister the device for tracking purpose or unregister custom assets like fonts this simply cant be done without a custom hook

its a necessity especially since MAUI based apps need msix to be used

**Comment by dongle-the-gadget** [comment]
> > https://task.ms/34863092
> 
> cant access

It's an internal bug tracker.

**Comment by untoldbyte** [comment]
> > > https://task.ms/34863092
> > 
> > cant access
> 
> It's an internal bug tracker.

Can we at least get an update

**Comment by stevenbrix** [comment]
>I'm very concerned this is being requested for a partner to pop open a web browser for some tracking, analytics, exit survey, and/or other generally unwanted behavior.

i'm not sure why an exit survey would be considered unwanted? i think it would be useful to know why people are uninstalling the app

**Comment by riverar** [comment]
@stevenbrix It could be a webpage that has privacy implications--could fingerprint the machine, browser usage, identify the user, IP address, etc. It's also disruptive to the user experience in general--what if the user is updating their browser? what if the browser is blocked and generates a weird policy error? what if the user doesn't have internet access?

**Comment by DrusTheAxe** [comment]
> @stevenbrix Privacy: The survey could fingerprint the machine, browser usage, identify the user, IP address, etc. General UX: What if the user is updating their browser? What if the browser is blocked and generates a weird policy error? What if the user doesn't have internet access?

MediumIL apps can already do much of this. This specific feature is support for a backgroundtask to be run on uninstall, just like on (pre)install and update. I'm not seeing how an uninstalllTask creates new threats. Aren't your privacy concerns aren't a factor for (pre)install and update?

**Comment by riverar** [comment]
@DrusTheAxe Agreed. I was just addressing Steven's general question about how exit surveys could be considered unwanted.

In this case, _updateTask_ appears to be unrestricted so the cat is already out of the bag. Completing the toolbox with uninstall, reset, repair, seems to make sense.

**Comment by JosHuybrighs** [comment]
Some uses cases where an Uninstall Task would be beneficial for me:

- An app needs to register entries in the `Windows Task Schedular` in order to allow the user to configure a given task to execute at some point in time without the app being open.
An Uninstall task is then needed to remove possible entries in the Task Schedular.
- An app needs a long living `Worker Service` (e.g. running as a `Windows Service`) to do some processing in the background without user interaction. It should start upon system boot and provides services that are not possible with InProc/OOP Background Tasks.
An Uninstall Task is needed to remove the service.

Note 1: The kind of work I am doing with the WinAppSDK is in the domain of file and folder backup/sync tools. Hence the above use cases.
Note 2: I have currently an app in the Microsoft Store that implements some form of Worker Service by means of a 2nd WPF process that is part of the MSIX package. It is registered to run upon boot (without a window) and then basically waits for events to occur and/or a Task.Delay to return. It does the job but it feels like not being the right way to accomplish what I want to do.

**Comment by DrusTheAxe** [comment]
> IMO the install and uninstall tasks should be available to any MSIX app, but then run with the trust level specified for the package. So, either lowIL/AppContainer, mediumIL ("full trust"), or even highIL. No restrictions about OEM, provisioning etc. please.

It sounds like you're asking for 2 things:
* windows.uninstallTask support, as the moral equivalent to 'uninstall' as [windows.updateTask](https://learn.microsoft.com/uwp/schemas/appxpackage/uapmanifestschema/element-1-extension) is to 'update'
* windows.*Task able to run with any supported TrustLevel (AppContainer, MediumIL, etc) just as any other activatable extension point

The 'Note' under [Remarks](https://learn.microsoft.com/uwp/schemas/appxpackage/uapmanifestschema/element-1-extension#remarks) imply the latter is true today. Is there a desired case not currently supported?


> (un)installation of drivers

This has been a recurring topic of discussion but there's complications. For example, drivers typically require admin privilege which most MSIX features don't require (installing a packaged containing a Service is one exception). There's also a wide variety of behavior across driver install/update/uninstall and some of these are at odds with core MSIX value prop (there's good ways to write driver installs that are well behaved citizens supporting clean install and uninstall, and many ways to be...less good citizens, with a lot of gray in between).

I'm not saying MSIX can't support drivers. Just that it's a harder problem to solve than most, and one of many things on the wishlist/backlog. Given infinite time and resources all is possible. Given not-so-unlimited real world constraints...

It's the constant juggle of priorities vs capacity. So please continue sharing what MSIX enhancements you'd like to see. Community feedback plays an important role in our planning. The more feedback we get the easier it is for us to prioritize the tasks you all would like to see.

**Comment by JosHuybrighs** [comment]
@DrusTheAxe 

> Why TaskScheduler? Why not a BackgroundTask?

That would work also but has the limitation that the minimum timeout is 15 minutes. If one also needs scheduled task execution at a given time of the day, week, .. one basically has to develop an own schedular service inside the task. Windows Task Schedular does all of this already.

> Sounds like a job for a service (not merely a process that performs service-y functionality), or what Unix calls a daemon or demon process. Does MSIX's support for services fit your need? If not then please share.

I have checked the use of a Packaged Service earlier and decided not to use it because of the following statement in [App capability declarations](https://learn.microsoft.com/en-us/windows/uwp/packaging/app-capability-declarations#restricted-capabilities).

> We don't recommend that you declare this capability in applications that you submit to the Microsoft Store. In most cases, the use of this capability won't be approved.

Isn't that a no-go then?

That is why I have chosen (or have been forced to choose?) to have a WPF process that runs at startup without an active window. I admit it provides a lot of things that are not so common, like:

- Periodic and scheduled task execution (file/folder  backup/sync).
- Task execution upon receiving USB storage device insert triggers.
- Task execution when detecting file changes.
- Handling file upload requests from trusted mobile devices using multicast SSDP and an encrypted upload service.

Probably not the most exiting features in 2024 for Microsoft (with OneDrive being promoted so heavily), but it seems there is still quite some interest in such services.

**Comment by whiskhub** [comment]
> The 'Note' under [Remarks](https://learn.microsoft.com/uwp/schemas/appxpackage/uapmanifestschema/element-1-extension#remarks) implies the latter is true today. Is there a desired case not currently supported?

The updateTask is working as expected, so no problems here. The `windows.preinstallConfigTask` is currently limited to provisioned/OEM apps, which is a bummer. And the uninstallTask does not exist yet. So my point was simply to make sure all will behave similarly and there are no artificial limitations except the trust level.

> I'm not saying MSIX can't support drivers. Just that it's a harder problem to solve than most, and one of many things on the wishlist/backlog. Given infinite time and resources all is possible. Given not-so-unlimited real world constraints...

I totally see that. But as there are still many use cases where a driver is required, and people will fall back to MSI. Just have a look what hardware manufacturers do to install drivers, it's really ugly. A clean MSIX solution would be much nicer. And packaged services seems to have been solved already. Maybe the solution is also to allow apps to move more code into userspace, but that's an even bigger topic :)

**Comment by DrusTheAxe** [comment]
Thank you for your thoughtful and detailed feedback. This is of great help in our planning and design discussions as we weigh the many asks on of time.

>So my point was simply to make sure all will behave similarly and there are no artificial limitations except the trust level.

Agreed on the desire for no artificial limitations, except maybe `preinstallConfigTask`. That's a more complex matter, and orthogonal to this topic. This issue is focused on addressing the gaps in uninstall/reset/repair. If you're interested in seeing the `preinstallConfigTask` constraints changed in some way please create a new issue with your feedback so it can be given its due consideration and discussion.

>But as there are still many use cases where a driver is required

I hear you. I didn't mean to imply it's of no import and we'll never do it (or even that you're the first to ask). Merely that it's non-trivial given the complex issues, so the effort and ROI haven't been high enough so far, relative to the many other asks on our time. I too would love to see this supported but Rome wasn't burned in a day... :P 

Please keep letting us know what you (and everyone else) need and desire. We'd love to tackle them all, it's just some things will take longer than others. Continue sharing your feedback to help us shape our priorities to best meet everyone's needs. And thank you. The community's input is very much appreciated.

**Comment by riverar** [comment]
This comes up a lot in the community and just want to point out generally all of this already exists and has existed in Windows 10 1903+. It's just currently artificially limited to partner use. I think it's a bit disingenuous to not mention this at all or pretend the ask is brand new.

References
* https://learn.microsoft.com/uwp/schemas/appxpackage/uapmanifestschema/element-desktop6-uninstallaction
* https://github.com/MicrosoftDocs/winrt-related/issues/194

**Comment by DrusTheAxe** [comment]
>It's just currently artificially limited to partner use.

This is incorrect. It's restricted to packages that are in MSIXVC format and have the rescap, but the former is the primary restriction. After all anyone can make a package with every rescap if they so choose, if they never submit to the Store.

Obligatory Reminder: `rescap` is a misnomer. It should have been more properly named `storecap` as the only enforcement blocking a developer from using one is during Store submission. Packages not submitted to the Store face no such constraint.

> I think it's a bit disingenuous to not mention this at all or pretend the ask is brand new.

No one ever said it's brand new in concept. Quite the opposite. But we also don't want to throw open the doors to 'custom actions' with the all merriment and mayhem those introduce.

We want to provide the moral equivalent of windows.updateTask for uninstall, reset and repair activity _without_ bringing along all the security and system integrity issues those 'CustomInstall' extensions bring to the table. We obviously know how as MSIX supports `windows.preinstallConfigTask` and `windows.updateTask`. The only question is priority / business justification to warrant funding the work (vs the zillion other things we're asked for).

As I've said before, if this (or any other) feature interests you the reader, please share your feedback. Community input helps shape our priorities so please let us know.

**Comment by zhuxb711** [comment]
@DrusTheAxe Any progress could share？

**Comment by ekalchev** [comment]
@DrusTheAxe
This is real problem for MSIX environment. Here our use case - since the desktop shortcuts are completely broken in MSIX, we use widely addopted workaround https://github.com/mjfusa/msix-create-desktop-shortcuts for creating .lnk files on first start.

The problem is that after uninstall there nothing that can cleanup the already broken shortcuts.

**Comment by DrusTheAxe** [comment]
> [@DrusTheAxe](https://github.com/DrusTheAxe) This is real problem for MSIX environment. Here our use case - since the desktop shortcuts are completely broken in MSIX, we use widely addopted workaround https://github.com/mjfusa/msix-create-desktop-shortcuts for creating .lnk files on first start.
> 
> The problem is that after uninstall there nothing that can cleanup the already broken shortcuts.

Understood.

Thanks for the feedback! Hearing this sort of information really does help with our planning and and getting resources committed.

@zhuxb711 No news to share yet. Will update here when there is. Hang tight (and keep that feedback coming :-)

**Comment by majaeger** [comment]
Just to echo a couple more use cases, our app also has some custom uninstall logic:

* Remove app-specific entries from cred manager
* Cleanup registry keys that are marked as unvirtualized in app manifest

**Comment by ekalchev** [comment]
We have three workarounds for detecting package uninstall and we are experimenting with them and they seems viable in the initial tests. You need a background process running.

1. https://learn.microsoft.com/bg-bg/uwp/api/windows.applicationmodel.packagecatalog.packageuninstalling

2. Handle WM_QUERYENDSESSION event and inside the handler poll for 2 sec the EventViewer Microsoft-Windows-AppXDeploymentServer/Operational for eventId 607 and DeploymentOperation=2 (this seems to be Remove). You need pollng because it seems there is latency when uninstall start and when the event appears in the log, may be you can use other methods to read the EventViewer that won't result such latency. It is important that this event is generated before you app is uninstalled and you can react to uninstallation. There is problems of course - user not having access to EventViewer and etc.

3. Register Task Scheduler task that triggers on EventViewer eventId 607. The task will be triggered on every package Uninstall and you have to filter for your package. The problem is you need to either run an .exe outside your install directory because it would already be gone, or use powershell script logic base64 encoded. The problem we couldn't solve here was the flashing cmd.exe window for fraction of a second on every uninstall.So far option1 look most promising but knowing microsoft there will be some catch or something won't be working.

**Comment by DrusTheAxe** [comment]
> * Cleanup registry keys that are marked as unvirtualized in app manifest

One of the quintessential reasons justifying windows.uninstallTask :-)

> * Remove app-specific entries from cred manager

Now that's interesting. Another good reason why windows.uninstallTask is righteous.

I wonder if MSIX could automagically cleanup cred manager entries. If the caller's packaged it would seem plausible. Will mention it to the team as an interesting vFuture enhancement.

**Comment by DrusTheAxe** [comment]
#1 is the recommended solution. That's not a workaround to detect uninstall so much as a feature.

But...you're trying to detect *your* app's package being uninstalled - to then trigger your cleanup code (quickly, before your app quits)?

**Comment by ekalchev** [comment]
> [#1](https://github.com/microsoft/WindowsAppSDK/pull/1) is the recommended solution. That's not a workaround to detect uninstall so much as a feature.
> 
> But...you're trying to detect _your_ app's package being uninstalled - to then trigger your cleanup code (quickly, before your app quits)?

Yes. My app has a separate background process running most of the time, which can react to the uninstall event and quickly clean up the desktop shortcuts.
However, this is essentially a race condition against the OS. It isn't suitable for cleanup operations that take a significant amount of time because Windows will forcefully terminate the process during uninstallation whenever it decides to. Furthermore, most apps don't maintain a persistent background process, so this workaround won't apply to them at all.
I really hope we get a proper windows.uninstallTask soon, because relying on a background process to catch the event before it gets killed is inherently unstable.

**Comment by elibroftw** [comment]
Our app is using the following architecture

- MSIX
  - Client/GUI app
  - Elevated Service -> Writes something important to %APPDATA%/AppName

We want to cleanup `%APPDATA%/AppName` on uninstall.

We actually don't need to use `%APPDATA%/AppName`, and are open to using the sandboxed directory to read/write the files, but it's unclear how to do that after reading the docs and this thread. Does anyone know how to do that?

```appxmanifest
<desktop6:Extension Category="windows.service" EntryPoint="Windows.FullTrustApplication" Executable="service.exe">
          <desktop6:Service Name="AppService" StartAccount="localSystem" StartupType="auto" Arguments="windows-service" />
        </desktop6:Extension>
```

Is there a way to specify the sandboxed dir through arguments?

Like right now, the service is installed in the same dir as the app is installed, I just don't know how to gain access to the redirect.

**Comment by elibroftw** [comment]
So there's a few solutions. This one seems to be the most robust imo.

```
Get-AppxPackage *AppName* | Select-Object PackageFamilyName
```

Now write config files to `%LOCALAPPDATA%\Packages\{PackageFamilyName}\AppData`

Unfortunatley, my service runs with StartAccount="localSystem" so I'm not too sure what happens

**Comment by DrusTheAxe** [comment]
>Writes something important to %APPDATA%/AppName

Does [ApplicationData's MachineFolder](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/applicationdata/ApplicationData.md#34-per-machine-data-store) fit your needs?

TL;DR 
1. Add `<appdata:ApplicationData><MachineFolder/></appdata:ApplicationData>` to your appxmanifest.xml
2. Access via [Microsoft.Windows.Storage.ApplicationData.MachinePath](https://learn.microsoft.com/windows/windows-app-sdk/api/winrt/microsoft.windows.storage.applicationdata.machinepath)

One caveat: Your packaged service runs as LocalSystem but MSIX can't register packages for LocalSystem* (packaged services are weird bending some of the normal rules). If you use WinAppSDK/SelfContained you have access to the M.W.S.ApplicationData API. If you use WinAppSDK/MSIX you don't, but it's a well-defined stable path so you could copy the code in [ApplicationData::_MachinePath(hstring const& packageFamilyName)](https://github.com/microsoft/WindowsAppSDK/blob/main/dev/ApplicationData/M.W.S.ApplicationData.cpp) line 292+, or do equivalent logic.

Key difference between `ApplicationData.MachinePath` and `%APPDATA%/AppName` is the former is automagically removed when the package is removed from the system, vs anything your service writes to %APPDATA%/AppName will hang around after you're uninstalled (desired or not).

\* Ideally you'd register the package as LocalSystem, use ApplicationData.LocalFolder and everything just works, but LocalSystem is the one account that *can't* register packages. Today. It's a topic of active discussion so I'm hopeful this issue will go away in the future, but until then... See above suggestions.

**Comment by elibroftw** [comment]
Okay so I think I confused myself. The config files I was writing to are not the same as when I'm running after installing from MSIX. I actually have no idea where `%APPDATA%\APPNAME` is pointing to for the service but I do know the files aren't being deleted on uninstall?

The service app is written in Rust, whereas the GUI app is C# self contained. Not sure if MachinePath still doable?

**Comment by DrusTheAxe** [comment]
>I actually have no idea where `%APPDATA%\APPNAME` is pointing to for the service

```
C:\Windows\System32> psexec -s c:\windows\system32\cmd.exe
...
C:\Windows\System32> ECHO %APPDATA%
C:\WINDOWS\system32\config\systemprofile\AppData\Roaming
```
:-)

>but I do know the files aren't being deleted on uninstall?

As expected. MSIX doesn't own or manage the classic unpackaged `%APPDATA%` location so MSIX won't do anything about it on package uninstall.
 
> The service app is written in Rust, whereas the GUI app is C# self contained. Not sure if MachinePath still doable?

Sure. C# and Rust support WinRT so you could call the M.W.S.ApplicationData.MachinePath API to get the path.

Or clone the code or do equivalent logic, if necessary. The important part is to declare MachineFolder in your appxmanifest.xml so MSIX creates the dir when the 1st package in the family enters the machine, and deletes it (and its content) when the last package in the family leaves the machine. That gets you the usual MSIX automagic create+delete support. No files left behind :-)

You can also adjust the MachineFolder's ACL to grant more or less access - depends if you want to, say, restrict access to only LocalSystem, grant write access to all users, or other games. See the `Sddl=...` attribute mentioned in the spec

https://github.com/microsoft/WindowsAppSDK/blob/main/specs/applicationdata/ApplicationData.md#34-per-machine-data-store

**Comment by elibroftw** [comment]
What is the best practice of getting the package family name from another binary? Or is there a better way to include binaries other than copying them in a prebuild phase?

**Comment by DrusTheAxe** [comment]
> What is the best practice of getting the package family name from another binary?

Sorry, I'm not following. Is this some process querying the package family name for a .msix file? A process query'ing package family name for _another_ running packaged process? Or something else?

**Comment by riverar** [comment]
> What is the best practice of getting the package family name from another binary? Or is there a better way to include binaries other than copying them in a prebuild phase?

1. handle = OpenProcess(PROCESS_QUERY_LIMITED_INFORMATION, ..., processId);
2. GetApplicationUserModelId(handle, ..., &aumidBuffer);
  > [!TIP]
  > Tip: You can allocate a MAX_AUMID_LEN (512) sized buffer to avoid calling GetApplicationUserModelId two times.

**Comment by elibroftw** [comment]
I'm getting a Windows cannot find error when I try to open in the explorer. Is that expected?

```
%ProgramData%\Microsoft\Windows\AppRepository\Families\ApplicationData\PACKAGE_FAMILY_ID\Machine
```

**Comment by elibroftw** [comment]
I got it working. thanks

**Comment by DrusTheAxe** [comment]
> > What is the best practice of getting the package family name from another binary? Or is there a better way to include binaries other than copying them in a prebuild phase?
> 
> 1. handle = OpenProcess(PROCESS_QUERY_LIMITED_INFORMATION, ..., processId);
> 2. GetApplicationUserModelId(handle, ..., &aumidBuffer);

Get... pkg/app identity APIs in appmodel.h tend to come in 3 flavors:
 
* GetCurrent* -- current process
* Get* -- process handle
* Get*FromToken -- token

> Tip: You can allocate a MAX_AUMID_LEN (512) sized buffer to avoid calling GetApplicationUserModelId two times.

+1000

You can (and should) preallocate fixed worst-case buffers:
 
* WCHAR packageFullName[PACKAGE_FULL_NAME_MAX_LENGTH + 1]{};
* WCHAR packageFamilyName[PACKAGE_FAMILY_NAME_MAX_LENGTH + 1]{};
* WCHAR applicationUserModelId[APPLICATION_USER_MODEL_ID_MAX_LENGTH]{};

**WARNING:** `PACKAGE_*_LENGTH` constants are the number of characters *not including* a null-terminator vs `APPLICATION_*_LENGTH` are the number of characters *including* a null-terminator. Thus the `+1` needed for package constants.

**NOTE:** `PACKAGE_RELATIVE_APPLICATION_ID_*_LENGTH` constants are number of characters *including* a null-terminator. Uncommonly used but keep in mind if you do.

>MAX_AUMID_LEN (512)

Did you mean `APPLICATION_USER_MODEL_ID_MAX_LENGTH`? That's 130

```
APPLICATION_USER_MODEL_ID_MAX_LENGTH
== PACKAGE_FAMILY_NAME_MAX_LENGTH + 1 + PACKAGE_RELATIVE_APPLICATION_ID_MAX_LENGTH
== PACKAGE_NAME_MAX_LENGTH + 1 + PACKAGE_PUBLISHERID_MAX_LENGTH 1 + PACKAGE_RELATIVE_APPLICATION_ID_MAX_LENGTH
== 50 + 1 + 13 + 1 + 64 + 1
== 130
```

**Comment by DrusTheAxe** [comment]
> I'm getting a Windows cannot find error when I try to open in the explorer. Is that expected?
> %ProgramData%\Microsoft\Windows\AppRepository\Families\ApplicationData\PACKAGE_FAMILY_ID\Machine


FYI that path won't exist before your package is registered

If memory serves...

`...\AppRepository` always exists on the system since Win8

`Families` and `ApplicationData` exist on new(ish) Windows images and created if not when the Machine folder's created

`PACKAGE_FAMILY_ID\Machine` is created potentially when the 1st package in the family is Staged, up to worst case when the 1st package in the family is registered for the first user

That last is a complicated story I plan to explain in an [Inside MSIX blog](https://devblogs.microsoft.com/insidemsix/) post
